api/bots/github_detail/github_detail.py

```diff
@@ -0,0 +1,116 @@
+import re
+import os
+import sys
+import logging
+import six.moves.configparser
+
+import requests
+
+class GithubHandler(object):
+    '''
+    This bot provides details on github issues and pull requests when they're
+    referenced in the chat.
+    '''
+
+    CONFIG_PATH = os.path.expanduser('~/github_detail.ini')
+    GITHUB_ISSUE_URL_TEMPLATE = 'https://api.github.com/repos/{owner}/{repo}/issues/{id}'
+    GITHUB_PULL_URL_TEMPLATE = 'https://api.github.com/repos/{owner}/{repo}/pulls/{id}'
+    HANDLE_MESSAGE_REGEX = re.compile("(?:([\w-]+)\/)?([\w-]+)?#(\d+)")
+    MAX_LENGTH_OF_MESSAGE = 200
+
+    def __init__(self):
+        config = six.moves.configparser.ConfigParser()
+        with open(self.CONFIG_PATH) as config_file:
+            config.readfp(config_file)
+        if config.get('GitHub', 'owner'):
+            self.owner = config.get('GitHub', 'owner')
+        else:
+            # Allowing undefined default repos would require multiple triage_message regexs.
+            # It's simpler to require them to be defined.
+            sys.exit('Default owner not defined')
+
+        if config.get('GitHub', 'repo'):
+            self.repo = config.get('GitHub', 'repo')
+        else:
+            sys.exit('Default repo not defined')
+
+    def usage(self):
+        # type: () -> None
+        return ("This plugin displays details on github issues and pull requests. "
+                "To reference an issue or pull request usename mention the bot then "
+                "anytime in the message type its id, for example:\n"
+                "@**Github detail** #3212 zulip/#3212 zulip/zulip#3212\n"
+                "The default owner is {} and the default repo is {}.".format(self.owner, self.repo))
+
+    def triage_message(self, message, bot_handler):
+        # type: () -> bool
+        # Check the message contains a username mention, an issue idi
+        # or 'help', and that we're not replying to another bot.
+        regex = "(?:@(?:\*\*){}).+?(?:#\d+)|(?:help)".format(re.escape(bot_handler.full_name))
+        return re.search(regex, message['content']) and not message['sender_email'].endswith('-bot@zulip.com')
+
+    def format_message(self, details):
+        # type: (Dict[Text, Union[Text, int, bool]]) -> Text
+        number = details['number']
+        title = details['title']
+        link = details['html_url']
+        # Truncate if longer than 200 characters.
+        ellipsis = '...'
+
+        if len(details['body']) > self.MAX_LENGTH_OF_MESSAGE + len(ellipsis):
```

Comment:
I don't think the logic here is right. If the message has a length of `MAX_LENGTH_OF_MESSAGE - 2`, it would get ellipsized (while it could be shown without problems).

Also, doing this makes that `MAX_LENGTH_OF_MESSAGE` is actually the maximum length - the length of the ellipsis.

I'd do something like:

```python
if len(details['body']) > self.MAX_LENGTH_OF_MESSAGE:
    description = details['body'][:self.MAX_LENGTH_OF_MESSAGE - len(ellipsis)] + ellipsis
```

(untested code, might be buggy)